lib/decorateWithMetrics.js

```diff
@@ -0,0 +1,132 @@
+import _ from 'underscore';
+
+/**
+ * Each key is a method name and the value is an array of calls metadata
+ * @type {Record<String, Array<CallMeta>>}
+ */
+let stats = {};
+
+/**
+ * Class instances are easily tracked by Dev tools memory snapshots
+ * Wrapping a result with such instance allows us to track it without
+ * incurring extra weight
+ */
+class TaggedItem {
+    constructor(data) {
+        Object.assign(this, data);
+        this.ref = data;
+    }
+}
+
+/**
+ * Capture tags in a weak map
+ * When the original object (key) is no longer used anywhere in app - the data (TaggedItem)
+ * is free for garbage collection and removed
+ * Using dev tools we see which TaggedItems are still retained and what's using them - preventing GC
+ * @type {WeakMap<object, TaggedItem>}
+ */
+const taggedItems = new WeakMap();
+
+/**
+ * Wraps a function with metrics capturing logic
+ * @param {function} func
+ * @param {String} [alias]
+ * @returns {function} The wrapped function
+ */
+function decorateWithMetrics(func, alias = func.name) {
+    if (_.has(stats, alias)) {
+        throw new Error(`"${alias}" is already decorated`);
+    }
+
+    stats[alias] = [];
+
+    function decorated(...args) {
+        const startTime = performance.now();
+        let result;
+
+        const originalReturnValue = func.apply(this, args);
+
+        /*
+        * Promisify the original return value in case the decorated method does not return a promise
+        * Then handlers added here are not affecting the returned promise as they aren't attached to it
```

Comment:
I think it would be more straightforward to skip timing here since there's nothing to time.

Reply:
Another option would be to throw when `originalReturnValue` is not returning a `Promise` and enforce that the decorated function return one.

Reply:
Oki, I just wanted to show that it's possible to make this more universal and work for sync methods as well

Reply:
It also covers the case where `merge` would not always return a promise - in case the Promise handling changes in `merge` get reverted

Reply:
>Oki, I just wanted to show that it's possible to make this more universal and work for sync methods as well

But we are not measuring any sync methods

>It also covers the case where merge would not always return a promise - in case the Promise handling changes in merge get reverted

If we are worried about behavior changing then we can write a test.

Reply:
> > Oki, I just wanted to show that it's possible to make this more universal and work for sync methods as well
> 
> But we are not measuring any sync methods?

We're not measuring any sync methods because:
1. Before this change it wasn't possible
2. You haven't said explicitly what you need tested

Now any sync method can be added and tested to see its effects on the bigger picture, if a sync method is taking more than 16ms it's a tell for a performance issue

Reply:
> > It also covers the case where merge would not always return a promise - in case the Promise handling changes in merge get reverted
> 
> If we are worried about behavior changing then we can write a test.

I was talking specifically for my current changes to `merge` if I revert them and the method is sometimes `void` sometimes returns a promise it won't be able to work without the above handling.
It's still not clear whether my changes to `merge` would stay or I need to find a different way to measure `merge` or not measure it at all

Reply:
>You haven't said explicitly what needs benchmarking

I'm just reviewing this PR 🤷‍♂️ 

But my assumption is that we were mainly concerned with benchmarking the performance of `AsyncStorage` so yeah, I'm confused by the discussion about async/sync methods.

Reply:
Ok, so we don't know exactly what needs benchmarking but this change allows to benchmark any method, how's that a bad thing?

Reply:
Left a comment, but I'd rather discuss how to move forward